File: src/chip8.zig
const cpu = @import("./cpu.zig");
const Cpu = cpu.Cpu;
const PROG_START = cpu.PROG_START;
pub const Register = cpu.Register;
pub const FRAME_SIZE = cpu.FRAME_SIZE;

pub const Chip8 = struct {
    cpu: Cpu = Cpu{},

    pub fn init() Chip8 {
        return .{};
    }

    pub fn load(self: *Chip8, bytes: []const u8) void {
        self.cpu.load(bytes);
    }

    pub fn update(self: *Chip8, delta: f32) !void {
        try self.cpu.update(delta);
    }

    pub fn step(self: *Chip8) !void {
        try self.cpu.step();
    }

    pub fn frame_ptr(self: *Chip8) *const [FRAME_SIZE]u8 {
        return &self.cpu.frame;
    }

    pub fn prog_ptr(self: *Chip8) [*]const u8 {
        const base_ptr: [*]const u8 = &self.cpu.ram;
        return base_ptr + PROG_START;
    }

    pub fn get_register(self: *Chip8, register: Register) u32 {
        return self.cpu.get_register(register);
    }

    pub fn set_register(self: *Chip8, register: Register, value: u32) void {
        self.cpu.set_register(register, value);
    }
};

File: src/chip8.js
let wasm;

export function init(callback) {
  WebAssembly.instantiateStreaming(fetch("./bin/chip8-vm.wasm"))
    .then((obj) => (wasm = obj.instance.exports))
    .then(() => callback());
}

export function get_register(register, index) {
  const map = {
    pc: wasm.pc,
    i: wasm.i,
    sp: wasm.sp,
    st: wasm.st,
    dt: wasm.dt,
    v: wasm.v,
  };
  if (index) {
    return map[register](index);
  }
  return map[register]();
}

export function set_register(register, index, value) {
  const map = {
    pc: wasm.set_pc,
    i: wasm.set_i,
    sp: wasm.set_sp,
    st: wasm.set_st,
    dt: wasm.set_dt,
    v: wasm.set_v,
  };
  if (index) {
    map[register](index, value);
  }
  map[register](value);
}

export function reset() {
  wasm.reset();
}

export function frame() {
  return new Uint8ClampedArray(
    wasm.memory.buffer,
    wasm.frame_ptr(),
    wasm.frame_size(),
  );
}

export function load(buffer) {
  let dest = new Uint8Array(
    wasm.memory.buffer,
    wasm.prog_ptr(),
    buffer.byteLength,
  );
  dest.set(new Uint8Array(buffer));
}

export function update(delta) {
  wasm.update(delta);
}

export function step() {
  wasm.step();
}
